Slugs & url improvements & Proof of Humanity [QC-143]

vidvidvid opened the pull request:
### URLs

URLs will be turned from something like:
`https://questchains.xyz/chain/0x89/0xeea5b8336aeaaa151043fd37a10c1d9b31f46171`
to
`https://questchains.xyz/polygon/musashis-meander`.

### Slugs

slug is auto generated while the quest chain is generated.
all the special characters are removed. spaces are turned into dashes:

`slug test-@#-bla-` will turn into `slug-test-bla`
if `slug-test-bla` already exists, a random string with the length of 5 will be attached to it at the end.

### Proof of Humanity

Profile page:

https://user-images.githubusercontent.com/35112939/209579643-057aa84c-cab4-497b-91f5-33422420d203.mov

Quest Chain view:

https://user-images.githubusercontent.com/35112939/209579634-fce315ba-4722-487d-9f16-22f5029cb932.mov

Closes #124 
Closes #140

vidvidvid commented:
@dan13ram please give it a look when you find some time. i'm also having a weird bug, which is multiplying instances of quest chains in the explore page & i couldn't figure out what happened yet. i tried
- reverting to earlier versions of the SDK
- checking out commit by commit to figure out when did i break it
- i tried wiping the node modules and reinstalling
& so far i found no clue. the explorer fetches quest chains normally when i checkout to an unrelated branch (like dev) could have we broken it when deploying the new subgraphs? any help would be appreciated a lot. 
![Screenshot 2022-12-22 at 23 55 11](https://user-images.githubusercontent.com/35112939/209238978-8a7332c6-8b59-4384-bc0c-18e7c4dd32bc.png)

vidvidvid commented:
weird, the deployed version has only 2 copies of each quest chain: https://deploy-preview-138--quest-chains-dev.netlify.app/explore

vidvidvid commented:
the subgraph has been fixed!

vidvidvid commented:
i've also removed the input from editor and qc creation wizard

dan13ram commented on components/CreateChain/MetadataForm.tsx:
```diff
@@ -22,34 +24,76 @@ import { isSupportedNetwork, useWallet } from '@/web3';
+        setSlug(slug);
+      } else {
+        setSlugAvailable(false);
+        setSlug(`${slug}${makeId()}`);
```
should we add a hyphen between the random id and the generated slug?

vidvidvid replied:
that's a good idea, i'll add it!

dan13ram commented on components/QuestChain/QuestChainV1Page.tsx:
```diff
@@ -65,6 +66,17 @@ import { UploadImageForm } from '../UploadImageForm';
 import { QuestsEditor } from './QuestsEditor';
 import { RolesEditor } from './RolesEditor';
 
+export const PoHAPI =
```
can we move these constants to a separate utils file?

vidvidvid replied:
done

dan13ram commented on components/QuestChain/QuestChainV1Page.tsx:
```diff
@@ -235,6 +254,35 @@ export const QuestChainV1Page: React.FC<QuestChainV1PageProps> = ({
+  const getPOH = async () => {
+    const statuses: { [addr: string]: boolean } = {};
+
+    Object.keys(members).forEach(async address => {
```
Could we query all members in a single api call instead of looping through them ?

vidvidvid replied:
done

dan13ram commented on components/QuestChain/QuestChainV1Page.tsx:
```diff
@@ -961,8 +1013,8 @@ export const QuestChainV1Page: React.FC<QuestChainV1PageProps> = ({
                         <NextLink
-                          as={`/chain/${questChain.chainId}/${questChain.address}/review`}
-                          href={`/chain/[chainId]/[address]/review`}
+                          as={`/${questChain.chainId}/${questChain.address}/review`}
```
```suggestion
                          as={`/${AVAILABLE_NETWORK_INFO[questChain.chainId].urlName}/${questChain.address}/review`}
```

dan13ram replied:
There are a few other places where we need to fix this i think

vidvidvid replied:
think i covered all the cases now. i just left the v0 quest chains

dan13ram commented on pages/[chainId]/[address]/index.tsx:
```diff
@@ -125,24 +131,45 @@ export async function getStaticPaths() {
 export const getStaticProps = async (
   context: GetStaticPropsContext<QueryParams>,
 ) => {
+  let chainId = context.params?.chainId;
```
should we rename the param `chainId` to `chainName` to be more clear ?

vidvidvid replied:
what about just `network`?

dan13ram replied:
yeah sure. it's just that `chainId` in the context of this codebase is used for the hex string of the chain number and we just need another name here

vidvidvid commented:
@dan13ram all of it should be fixed now

dan13ram commented on pages/profile/[address].tsx:
```diff
@@ -20,8 +32,25 @@ const Profile: React.FC<Props> = ({ address: addressURL }) => {
+
+  useEffect(() => {
+    getPOH();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
we should either put the lines inside `getPOH` method into this `useEffect` or wrap the `getPOH` method in a useCallback

dan13ram replied:
We could also create a simple hook that takes the `address` and returns a `registered` boolean.

dan13ram commented on components/QuestChain/QuestChainV0Page.tsx:
```diff
@@ -701,8 +701,8 @@ export const QuestChainV0Page: React.FC<QuestChainV0PageProps> = ({
-                        as={`/chain/${questChain.chainId}/${questChain.address}/review`}
-                        href={`/chain/[chainId]/[address]/review`}
+                        as={`/${questChain.chainId}/${questChain.address}/review`}
+                        href={`/[chainId]/[address]/review`}
```
```suggestion
                        href={`/[network]/[address]/review`}
```

The `[chainId]` page no longer exists right? For both v0 and v1 its the same url, isn't it?

vidvidvid replied:
true

dan13ram commented on components/ProfileView/QuestsToReview.tsx:
```diff
@@ -86,8 +86,10 @@ const QuestChainStatusView: React.FC<{
+    as={`/${AVAILABLE_NETWORK_INFO[chain.chainId].urlName}/${
+      chain.address
+    }/review`}
+    href={`/[chainId]/[address]/review`}
```
```suggestion
    href={`/[network]/[address]/review`}
```

Need to change this wherever applicable.

dan13ram commented on components/ProfileView/NFTsToMint.tsx:
```diff
@@ -78,7 +78,7 @@ export const NFTsToMint: React.FC = () => {
               {nftsToMint.slice(0, 2).map(ns => (
                 <MintNFTTile
-                  QCURL={`${QUESTCHAINS_URL}/chain/${ns.questChain.chainId}/${ns.questChain.address}`}
+                  QCURL={`${QUESTCHAINS_URL}/${ns.questChain.chainId}/${ns.questChain.address}`}
```
shouldn't this be the network name instead of chainId ?

vidvidvid replied:
i missed that, thanks!

dan13ram commented on components/CreateChain/MetadataForm.tsx:
```diff
@@ -22,34 +24,76 @@ import { isSupportedNetwork, useWallet } from '@/web3';
 
   const [isSubmitting, setSubmitting] = useState(false);
 
+  const fetchSearchResults = async (slug: string) => {
```
should wrap this in a useCallback

dan13ram commented on .eslintrc.json:
```diff
@@ -23,7 +23,6 @@
     "react/display-name": "off",
     "react/no-unescaped-entities": "warn",
     "react-hooks/rules-of-hooks": "error",
-    "react-hooks/exhaustive-deps": "error",
```
this is an important react/eslint rule that shouldn't be taken off imho.

vidvidvid replied:
"react-hooks/exhaustive-deps" was set twice in that file, once with "off" and once with "error". tbh sometimes i found it annoying, but it might be better to not completely ditch it. maybe make it into a warning instead?

dan13ram replied:
sure warning should be fine. 
we should ensure to fix these lint warnings before merging though, because this lint rule helps us from some unwanted bugs that may prop up with not having the correct dependencies or having unwanted ones.

dan13ram replied:
@vidvidvid https://dev.to/wkrueger/never-ignore-the-exhaustive-deps-rule-2ap8

vidvidvid replied:
let's put the error back in then :)

dan13ram commented on pages/[network]/[address]/review.tsx:
```diff
@@ -152,23 +156,44 @@ export async function getStaticPaths() {
-      );
+
+  if (network && CHAIN_URL_MAPPINGS[network]) {
+    network = CHAIN_URL_MAPPINGS[network];
```
shouldn't we return null or fail if `CHAIN_URL_MAPPINGS[network]` is not found? 
or we should find a way to redirect to the new networkName url when the user is inputting the hexstring chainId

vidvidvid replied:
redirecting sounds like a good idea, yeah. but i mainly put this code there to safeguard urls that will have the network ids instead of names, like this: `https://questchains.xyz/chain/0x89/0x0bf17bedbd42f6cf89f4064d5d70da1e44f70f14`

vidvidvid commented:
@dan13ram pls give it another 👀 
- i extracted the PoH functionality by separating single address & multiple into 2 hooks to make everything clearer
- reverted the `exhaustive-deps`
- cleared up a couple of other cases with routes
- redirecting all old routes to new ones
